src/algebra/module/polynomial.lean

```diff
@@ -0,0 +1,159 @@
+        { rw [pow_succ, mul_eq_comp, ← linear_map.comp_assoc, h, linear_map.comp_assoc, ih,
+            ← linear_map.comp_assoc, ← mul_eq_comp, ← pow_succ] } },
+      induction u using module_with_End.rec,
+      rw [comp_apply, comp_apply, coe_to_linear_map, coe_to_linear_map, smul_def, aeval_monomial,
+        mul_apply, algebra_map_End_apply, symm_apply_apply, L.map_smul,
+        comp_apply, comp_apply, coe_to_linear_map, coe_to_linear_map, smul_def, aeval_monomial,
+        mul_apply, algebra_map_End_apply, symm_apply_apply, ← comp_apply, ← comp_apply, this] }
+  end,
+  ..(of_module F).to_linear_map.comp $ L.comp $ (of_module E).symm.to_linear_map }
+@[simp] lemma polynomial_linear_of_semiconj_apply {L : M →ₗ[R] N} (h : L.comp E = F.comp L)
+  (u : M) : polynomial_linear_of_semiconj h (of_module E u) = of_module F (L u) := rfl
+
+/--A `R[X]`-linear isomorphism is a `R`-linear isomorphism compatible with the action of `X`.-/
+def polynomial_lequiv_of_semiconj {L : M ≃ₗ[R] N}
+  (h : (L : M →ₗ[R] N).comp E = F.comp (L : M →ₗ[R] N)) :
+  module_with_End E ≃ₗ[R[X]] module_with_End F :=
+{ map_smul' := (polynomial_linear_of_semiconj h).map_smul',
+  ..(of_module E).symm.trans $ L.trans (of_module F) }
+@[simp] lemma polynomial_lequiv_of_semiconj_apply {L : M ≃ₗ[R] N}
+  (h : (L : M →ₗ[R] N).comp E = F.comp (L : M →ₗ[R] N)) (u : M) :
+  polynomial_lequiv_of_semiconj h u = L u := rfl
+
+/--A `R[X]`-linear map is a `R`-linear map compatible with the action of `X`.-/
+noncomputable def semiconj_of_polynomial_linear (L : module_with_End E →ₗ[R[X]] module_with_End F) :
+  { L : M →ₗ[R] N // L.comp E = F.comp L } :=
+⟨(of_module F).symm.to_linear_map.comp $ (L.restrict_scalars R).comp (of_module E).to_linear_map,
+begin
+  ext u,
+  simp only [comp_apply, coe_to_linear_map, linear_map.restrict_scalars_apply],
+  rw [← X_smul, L.map_smul, symm_apply_eq, ← X_smul, apply_symm_apply]
+end⟩
+@[simp] lemma semiconj_of_polynomial_linear_apply (L : module_with_End E →ₗ[R[X]] module_with_End F)
+  (u : M) : of_module F (semiconj_of_polynomial_linear L u) = L (of_module E u) := rfl
+
+/--A `R[X]`-linear isomorphism is a `R`-linear isomorphism compatible with the action of `X`.-/
+noncomputable def semiconj_of_polynomial_lequiv (L : module_with_End E ≃ₗ[R[X]] module_with_End F) :
+  { L : M ≃ₗ[R] N // (L : M →ₗ[R] N).comp E = F.comp (L : M →ₗ[R] N) } :=
+⟨(of_module E).trans $ (L.restrict_scalars R).trans (of_module F).symm,
+begin
+  ext u,
+  simp only [comp_apply, linear_equiv.coe_coe, trans_apply, linear_equiv.restrict_scalars_apply],
+  rw [← X_smul, L.map_smul, symm_apply_eq, ← X_smul, apply_symm_apply]
+end⟩
+@[simp] lemma semiconj_of_polynomial_lequiv_apply (L : module_with_End E ≃ₗ[R[X]] module_with_End F)
+  (u : M) : of_module F (semiconj_of_polynomial_lequiv L u) = L (of_module E u) := rfl
+
+/--Equivalence between `R[X]`-linear maps and `R`-linear maps compatible with the action of `X`.-/
+noncomputable def polynomial_linear_equiv_semiconj :
```

Comment:
```suggestion
@[simps] noncomputable def polynomial_linear_equiv_semiconj :
```
to add the lemmas that the equiv is built out of `semiconj_of_polynomial_linear` and `polynomial_linear_of_semiconj`